### ISSUE TITLE
Scaling not updated upon return from fullscreen in pixel-perfect mode

### ISSUE THREAD

**Issue by ant-222** [dosbox-staging/dosbox-staging#808]
Hell, all

A neighbor has asked me today to help her with the computer (*Windows 10*), and I took the opportunity to show her son some good old games. I naturally chose your project as the only one that supports pixel-perfect mode (although I hope to get it in *DOSBox-X*, too), but neither *openglpp* nor *texturepp* worked well :-( *DOSBox* started well, entered full screen mode without a problem, but when I pressed `ALT+Enter` again to return to windowed mode, the scale factors were not decreased,  so I got a small window displaying the top-left corner of the larger image. I fear you broke something in the implementation of the patch. Can you please check it?

**Comment by kcgen** [comment]
Welcome (back) @ant-222 !
Thank you for the report, will indeed check it out.

**Comment by ant-222** [comment]
Thank you, @kcgen. Basically, you need only start some 320x200 game in windowed mode, and press `ALT+Enter` twice.

**Comment by ant-222** [comment]
Any good news? Feel free to ask me for additional experiments. I can do them at my friend's.

**Comment by kcgen** [comment]
Nothing yet, but starting to look into it now.
Will let you know ASAP when there's something to try.

**Comment by ant-222** [comment]
OK, I just thought I should make sure. Take your time. I am ready to help you by testing *dosbox-staging* at my various friends with new computers.

**Comment by kcgen** [comment]
Couple tests on Windows 10, setting `output = openglpp`, and all other `[sdl]` settings default, and was able to flip-flop between fullscreen and window-mode without problems.

![2021-01-09_15-40](https://user-images.githubusercontent.com/1557255/104110909-29cd7b80-5291-11eb-9240-2c4136f4b7aa.png)
![2021-01-09_15-41](https://user-images.githubusercontent.com/1557255/104110910-2b973f00-5291-11eb-9ff6-376a7f5d83dd.png)

Do you have a `conf` files and specific version that it might be happening on?

Alternatively, give the letest `master` build a shot too! - https://dosbox-staging.github.io/downloads/devel/

**Comment by ant-222** [comment]
Thanks for testing, @kcgen . As I said, I used the version downloaded from the official website on that very day. I generated the config file via

    config -writeconf

and changed only the *output* setting. I will try again at my friend's and, if the error occurs, will give you all the details: version and config. I will try the test build, too.

I do not the a report of the *available area* in the log—an important indicator of what area pixel-perfect mode thinks is available for magnfication. Just the other day it helped me find a problem with the handling of high-DPI awareness on high-DPI decies: it saw virtual pixels rather than physcial ones.

Notice, *Doom* does not look pixel-perfect on neither of you screen-shots. On the first one, the image should consist of 4x3 pixel blocks, but it does not. On the second—it should consist of 2x2 pixel blocks, but it does not. Can you please look into that while I re-test switching between windowed and full-screen modes at my friend's? Shall I provide magnified fragments of your screenshot that may see the problem clearer?

The size of the full-screen screenshot is 1546x927 and of the windowed one—1289x879. Why? Neither of them looks like a standard display resolution...

**Comment by kcgen** [comment]
I agree @ant-222.

I started to go through `sdlmain.cpp`, however it has image sizes, window sizes, and aspect handling in an excessive number of places, and state is also held and assigned across structs and local variables.

I'd like to wait until @dreamer can weigh in on how to address this, as he's made a couple corner-case fixes during prior months.  I know testing all of the resolution-mode handling is quite a chore given the number of architectures and output modes supported. 

That said, obviously there's a bug - and on Windows no less, so it's not a corner care or rare architecture.

**Comment by ant-222** [comment]
I have come to my friend and am typing from his PC with *Windows 7*. We tested the release version 0.76.0—and the bug is there. [Here](https://drive.google.com/file/d/1e-VtszwsKjP2pgY03m1YnMkKOCpJHha-/view?usp=sharing) is what it looks like after pressing `Alt+Enter` twice. I launched *DOSBox* with [this](https://pastebin.com/raw/nt257SSx) config file.

> Alternatively, give the letest `master` build a shot too! - https://dosbox-staging.github.io/downloads/devel/

We observe exactly the same bug with development build `v0.77.0-alpha-155-gd6ed`, so I need not attach a screenshot of it. The config file is [here](https://pastebin.com/raw/FdMsVK8D).

To summarise: I have now two cases—one on *Windows 7* and one on *Windows 10*—where this bug is present. Have you an idea why your builds may crash on my *Windows XP*?

**Comment by kcgen** [comment]
> Have you an idea why your builds may crash on my Windows XP?

Is this a separate issue?

**Comment by ant-222** [comment]
Sure it is a separate issue, or even a non-issue if *DOSBox Staging* does not claim support of *Windows XP*. If it does, though, I will create a separate issue.

**Comment by ant-222** [comment]
@kcgen:

> I'd like to wait until @dreamer can weigh in on how to address this, as he's made a couple corner-case fixes during prior months. I know testing all of the resolution-mode handling is quite a chore given the number of architectures and output modes supported.

In the meantime (but not mean time :-) we can perform a series of regression tests—perhaps by way of binary search between the initial pixel-perfect release the latest version—to determine the commit the broke it. I have no time right now to help you with the code, but I can assist with testing. Do I understand you right that you have not been able to reproduce it on your end?

**Comment by kcgen** [comment]
Hi @ant-222 , that's correct; and sounds like a good plan 👍

On Linux hardware running X11-XFCE, macOS Catalina VM, and latest Windows (64-bit) VM, I can toggle between full and window-mode repeatedly. The window (and its content) shrinks back to its original windowed-size, as one would expect.  So in this regard, I'm unfortunately not able to reproduce it.

That said - I do see the lack of aspect correction in the windowed modes - so this is something I can atleast work backward and bisect (and I suspect the issue is on our side). 

Regarding going back to prior releases - we had a similar report of full-screen toggle problems affecting *all* releases.  Well.. turns out the issue is with gnome 3.81; an no amount of fixing on our side was going to change it. 

Looking forward to what you find. Hopefully it's something we can fix, and not a new Window 10 version impacting SDL2.

**Comment by ant-222** [comment]
@kcgen

> I'm unfortunately not able to reproduce it.
> 
> That said - I do see the lack of aspect correction in the windowed modes - so this is something I can atleast work backward and bisect (and I suspect the issue is on our side).

Beware that pixel-perfect may not always be able to correct the aspect ratio. The printing of the available area, which I think you removed from the debug output window, would help a lot: what if the algorithm does not "see" sufficient elbowroom?

> Looking forward to what you find. Hopefully it's something we can fix, and not a new Window 10 version impacting SDL2.

As I said, I will need the actual builds, with which to go my friend for testing. I cannot currently produce them myself. Let us start maybe from the very first commit with pixel-perfect mode. If the bug is already there, then we may conclude it was broken from the start—I can only confirm it worked in my branch before the PR.

To summarise our problems—
1. failure, in some cases, to update scaling upon return from full-screen mode on *Windows 7* and *Windows 10* on [my side](https://github.com/dosbox-staging/dosbox-staging/issues/808#issuecomment-757522759),
2. [lack](https://github.com/dosbox-staging/dosbox-staging/issues/808#issuecomment-758231755) of aspect-ratio correction in window mode,
3. Incorrect scaling in the [screenshots](#issuecomment-757384841) that you attached: a pixel-level micropattern.

**Comment by ant-222** [comment]
This is a friendly reminder that I ready, willing, and able to participate in regression testing with the goal of pinning down the breaking change. All I need is a series of *Windows* builds to try on the affected machines.

**Comment by dreamer** [comment]
@ant-222 You can download builds going back many months back from here: https://github.com/dosbox-staging/dosbox-staging/actions?page=4&query=workflow%3A%22Windows+builds%22+branch%3Amaster

**Comment by ant-222** [comment]
Thank you, @dreamer. Hopefully I will start testing over this weekend.

**Comment by ant-222** [comment]
I am sorry, but the initial versions with pixel-perfect mode in March of 2020 are reported expired. It is possible somehow to re-order those historical from the *CI*? In the meantime, I have downloaded the oldest available *Windows* build #4895 and will test it.

**Comment by ant-222** [comment]
I have tested windows builds ## 4895 and 6078, both x86 and x64, and the bug is present in both. I don't think there is any sense in testing anything in between, whereas I have not been able to download any older builds for further testing.

**Comment by GranMinigun** [comment]
@ant-222, when you have a chance, please test the latest version. Setting viewport should be more robust now, hopefully solving the problem on your side.

**Comment by johnnovak** [comment]
@ant-222 Could you please test this with our latest release?

**Comment by Burrito78** [comment]
No he couldn't. Closing. PP is killed anyway.

**Comment by Marat-Tanalin** [comment]
> PP is killed anyway.

What is this supposed to mean?

**Comment by Burrito78** [comment]
> > PP is killed anyway.
> 
> What is this supposed to mean?

#2448

**Comment by johnnovak** [comment]
> > > PP is killed anyway.
> > 
> > 
> > What is this supposed to mean?
> 
> #2448

Good riddance!
